geopandas/io/geoarrow.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import pyarrow as pa
+
+import shapely
+from shapely import GeometryType
+
+
+def _get_inner_coords(arr):
+    if pa.types.is_struct(arr.type):
+        if arr.type.num_fields == 2:
+            coords = np.column_stack(
+                [np.asarray(arr.field("x")), np.asarray(arr.field("y"))]
+            )
+        else:
+            coords = np.column_stack(
+                [
+                    np.asarray(arr.field("x")),
+                    np.asarray(arr.field("y")),
+                    np.asarray(arr.field("z")),
+                ]
+            )
+        return coords
+    else:
+        # fixed size list
+        return np.asarray(arr.values).reshape(len(arr), -1)
```

Comment:
Is it easy to generate an input file to read so our tests cover the interleaved coordinates case? (we also miss coverage above for the 2D/3D case, but the logical is identical so I don't think that matters.

Reply:
So officially GeoParquet doesn't support interleaved coordinates, only separated. That's why it is not included here in the tests. In a follow-up PR, I however want to add support for converting from (Geo)Arrow (in-memory) to geopandas (https://github.com/geopandas/geopandas/pull/3219 is already doing the export from geopandas to geoarrow), and those code paths will be covered. 

Not ideal that for now it's not covered (and I could also have just included the separated code path here), but will do that follow-up shortly.

Reply:
For reference, this was done in #3301